### PR TITLE
Modified patient id creation

### PR DESCRIPTION
**Pull request by wynnliam**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Update check-in for patients.

* **What is the current behavior?** (You can also link to an open issue here)
Uses first name, last name, and last 4 digits of social security to create a patient id. All these fields are required.

* **What is the new behavior (if this is a feature change)?**
User can supply first name, last name, and email if they want. They are required to put in a 4 digit code. If they choose to add an email, it must be formatted properly.

* **Other information**:

### CODE REVIEW

File: src/app/check-in/check-in.component.ts [hunk 1]
```diff
@@ -13,96 +13,96 @@ import { ThsDataService } from '../services/ths-data.service';
...
+  }
 
-   /**
-    * This function will be call when the "check in" button is pressed.
-    * It evaluates the value is typed in the input area; if the value has the length of 6,
-    * then the page will be route to the "audiologist" page;
-    * if the value of length 4, then the page will be route to the "appointment" page.
-    * If the value has different length, then it will be informed as incorrect check in.
-    */
+  isPatientIdValid() : boolean {
```
Comment (ElijahRW) [inline src/app/check-in/check-in.component.ts:79]: You can actually use Regular expression to your advantage here: `isPatientIdValid` and `isPatientIdNumber` is all equivalent to the following code:

```  javascript
isPatientIdValid(): boolean {
    const PATIENT_ID_LENGTH = 4;
    let reg = new RegExp('^\\d{'+ PATIENT_ID_LENGTH +'}$');
    return (reg.test(this.patientId)) ? true : false;
  }
```

File: src/app/check-in/check-in.component.ts [hunk 2]
```diff
@@ -13,96 +13,96 @@ import { ThsDataService } from '../services/ths-data.service';
...
+    // Email is an optional field the user does not need
+    // to fill out. However, should they choose to do so, it
+    // should follow a proper format.
+    if(this.email.length === 0)
+      return true;
 
-      if (snd < 7) {
-        snd += 2;
-      } else { snd -= 3; }
+    let i: number;
```
Comment (ElijahRW) [inline src/app/check-in/check-in.component.ts]: To simplify this for loop: consider a slightly more complex regular expression
``` javascript
// Taken from: https://stackoverflow.com/questions/46155/how-to-validate-an-email-address-in-javascript
    var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return re.test(String(this.email).toLowerCase());
```

File: src/app/check-in/check-in.component.ts [hunk 3]
```diff
@@ -13,96 +13,96 @@ import { ThsDataService } from '../services/ths-data.service';
...
+    if(this.isCredentialsValid()) {
+      sessionStorage.clear();
+      this.tsDataService.clearHistory();
+      this.tfiDataService.clearHistory();
+      this.thsDataService.clearHistory();
+      Utilities.setSessionStorage('patient-id', this.patientId);
+      Utilities.setSessionStorage('first-name', this.firstName);
+      Utilities.setSessionStorage('last-name', this.lastName);
+      Utilities.setSessionStorage('email', this.email);
+      console.log(this.patientId);
```
Comment (ElijahRW) [inline src/app/check-in/check-in.component.ts]: remove `console.log` calls

File: src/app/check-in/check-in.component.html [hunk 4]
```diff
@@ -25,28 +25,31 @@
...
     <div class="col-sm-8 col-md-6 col-lg-4 col-lg-offset-4 col-md-offset-3 col-sm-offset-2">
-      <label><input id="lastfourssn" type="text" autofocus [(ngModel)]='lastFourSS' placeholder="Last 4 digits of SSN" > </label>
+      <label><input id="usercode" type="text" autofocus [(ngModel)]='patientId' placeholder="Usercode" > </label>
     </div>
+
   </div>
+
     <div class="row">
       <div class="col-sm-8 col-md-6 col-lg-4 col-lg-offset-4 col-md-offset-3 col-sm-offset-2">
         <div *ngIf="!authenticationFlag" class="failLogIn">One or more of the above fields is incomplete.</div>
```
Comment (ElijahRW) [inline src/app/check-in/check-in.component.html:55]: Point Error message to a component variable held here to be an error message to tell the user what rule they are violating:
```suggestion
        <div *ngIf="!authenticationFlag" class="failLogIn">{{errorMessage}}</div>
```

**Review by ElijahRW — CHANGES_REQUESTED** [review]
The error message needs to be more specific as to why a client input has been rejected.

But, looks good!

File: src/app/check-in/check-in.component.html (hunk 4 above)
Comment (movshov) [reply src/app/check-in/check-in.component.html:55]: I agree with Elijah's suggestion of holding the error variable.

**Review by movshov — APPROVED** [review]
Once you implement Elijah's suggestions it looks good to me. No obvious issues that i can see.

File: src/app/check-in/check-in.component.ts [hunk 5]
```diff
@@ -13,96 +13,90 @@ import { ThsDataService } from '../services/ths-data.service';
...
-    */
+    // Email is an optional field the user does not need
+    // to fill out. However, should they choose to do so, it
+    // should follow a proper format.
+    if(this.email.length === 0)
+      return true;
 
-   public onClick() {
+    // Credit for this here: https://tylermcginnis.com/validate-email-address-javascript/
+    // Doesn't seem to handle multiple @ symbols however.
```
Comment (gstone271) [inline src/app/check-in/check-in.component.ts]: Multiple @ symbols are not allowed to appear in an email.

**Pull request closed without merging**